[SYCL][Joint Matrix Spec] Add new API for out of bounds fill/load/store (#11172)
intel/llvm: sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_intel_matrix.asciidoc

   sum_local_rows[row] += val;
});
```

=== Load/Store/Fill With Out-of-Bounds Checks
`joint_matrix_load`, `joint_matrix_store`, and `joint_matrix_fill`

> operations do not do bounds checking. When the global matrix size does
> not multiply `Rows` and `Cols` of the joint matrix, this extension

Reviewer: ```suggestion
operations do not do bounds checking. When the global matrix size is
not multiple of the joint matrix size, this extension
```

adds a new form of API to load, store, and fill joint matrix while
checking the bounds. For load and fill, the out-of-bounds elements are
set to 0. For the store, they are dropped out.

> These functions are similar to the existing ones but they

Reviewer: ```suggestion
These functions are similar to the ones without bounds checking, but they
```

`Height` and `Width` of the two-dimensional region to load to or store
from.

The two-dimensional memory region is interpreted as follows, where
`Height` is number of rows in the memory matrix, `Width` is number of
columns in the memory matrix, `stride` is number of rows that include
the out of bounds data (depicted as x here).

|<-------------`Width`---------->|

|DDDDDDDDDDDDDDDDDxxx|

|DDDDDDDDDDDDDDDDDxxx|

|......................................................| `Height`

|DDDDDDDDDDDDDDDDDxxx|

|DDDDDDDDDDDDDDDDDxxx|

|<------------`Stride`---------->|
 
Note that joint matrix shape `Rows` and `Cols` represents a sub-block
of the picture above for each sub-group execution. At the last
iteration of the range of the launched kernel, the remainder that
should exclude the out-of-bounds data is checked against `Width` and
`Height`. For instance, load of `sub_b` defined as
`joint_matrix<sub_group, bfloat16, use::b, 2, 3, layout::row_major> sub_b;`
checks that `Height=10` is not multiple of `3` and the data (2
elements left) that exceeds getting the total `10` rows will be filled
by zero.

```c++
namespace sycl::ext::intel::experimental::matrix {

> 

Author: @fda0 does this change address the missing arguments to generate efficiently 2d load/store in IGC?

Author: Yes I changed it to add the two offset coordinates separately and excluded them from the base pointer calculation.

`col` represent the coordinates in the logical matrix before VNNI
transformation.

=== Load/Store/Fill With Out-of-Bounds Checks
`joint_matrix_load`, `joint_matrix_store`, and `joint_matrix_fill`
operations do not perform bounds checking. When the global matrix size is
not multiple of the joint matrix size, this extension adds a new form
of API to load, store, and fill joint matrix while checking the
bounds. For load and fill, the out-of-bounds elements are set to
0. For the store, they are dropped out.

These functions are similar to the existing ones without bounds
checking, namely `joint_matrix_fill`, `joint_matrix_load`, and
`joint_matrix_store`. But they are different in three ways:

* The pointer `base_src` or `base_dest` designates the base pointer of

> the original memory matrix excluding any offset calculations. The

Author: Yes, I like this wording better, thanks!

coordinates are given separately here.
* The coordinates `CoordX` and `CoordY` to calculate the  pointer

> offset to load/store are given as separate arguments.

Author: Correct

from.

The two-dimensional memory region is interpreted as follows, where
`Height` is number of rows in the memory matrix, `Width` is number of
columns in the memory matrix, `stride` is number of rows that include
the out of bounds data (depicted as x here).

|<-----------`Width`-------->|

|DDDDDDDDDDDDDDDDDxxx|

|DDDDDDDDDDDDDDDDDxxx|

|......................................................| `Height`

|DDDDDDDDDDDDDDDDDxxx|

|DDDDDDDDDDDDDDDDDxxx|

|<------------`Stride`---------->|

Note that joint matrix shape `Rows` and `Cols` represents a sub-block
of the picture above for each sub-group execution. The out of bounds
functions checks that the 2d block starting at the address to be
accessed `[0, src]` and ending at `[Cols, src + Rows * stride]` does
not exceed the bigger 2d block starting at `[0, src]` and ending at
`[Width, src + Height * stride]`. For instance, load of `sub_b`
defined as `joint_matrix<sub_group, bfloat16, use::b, 2, 3,
layout::row_major> sub_b;` from `ptr + 9` checks that `ptr + 9` does
not exceed `ptr + 10` where 10 is `Width` so it reads only 1 element
for that iteration. The 2 columns left that exceed the total `10`

> columns are filled by zero.

Reviewer 2: I find this paragraph extremely confusing.  Some specific things that seem confusing:

* The phrase "for each sub-group execution" seems irrelevant.  You are describing a single joint matrix load operation here, so there is no need to talk about iterations in a larger algorithm.
* What are `src` and `ptr`?  None of the APIs have parameters with these names.
* It seems like `CoordX` and `CoordY` should appear somewhere in this description since they indicate the starting point of the joint matrix.

```c++
namespace sycl::ext::intel::experimental::matrix {

template <typename Group, typename T, size_t Rows, size_t Cols,
          use Use, layout Layout, typename Tv>
void joint_matrix_fill_checked(Group g, joint_matrix<Group, T, Use, Rows,
          Cols, Layout> &m, Tv v, size_t Stride, size_t Height, size_t

>           Width, size_t coordX, size_t coordY);

Author: @MrSidims, 
In the load/store, the check does the following (let's just consider one horizontal dimension to simplify):
if (base_ptr + coordY + Cols > base_ptr+Width) --> out of bounds

Since we intend  to do the check just like load and store (just considering base_ptr=0), we need to add these arguments: Height, Width, Stride, CoordX, and CoordY. The check will do the following:
if (coordY + Cols > Width) --> out of bounds

template <typename Group, typename T, size_t Rows, size_t Cols,
          typename PropertyListT>
void joint_matrix_store_checked(Group g,
   const joint_matrix<Group, T, use::accumulator, Rows, Cols, layout::dynamic> &res, 
   ext::oneapi::experimental::annotated_ptr<T, PropertyListT> base_dest,
   size_t Stride, layout Layout, size_t Height, size_t Width,
   size_t coordX, size_t coordY);

// Only available when Layout != layout::dynamic
template <typename Group, typename T, size_t Rows, size_t Cols,
          use Use, layout Layout, typename PropertyListT>
void joint_matrix_store_checked(Group g,
   const joint_matrix<Group, T, Use, Rows, Cols, Layout> &res,
   ext::oneapi::experimental::annotated_ptr<T2, PropertyListT> base_dest,
   size_t Stride, size_t Height, size_t Width, size_t coordX, size_t coordY);

} // namespace sycl::ext::intel::experimental::matrix
```

The property list associated with the `annotated_ptr` argument
represents the compile-time constant properties for cache control included
in the SYCL extenion
link:../../proposed/sycl_ext_intel_cache_controls.asciidoc[sycl_ext_intel_cache_controls],
and for the alignment property defined in the SYCL extension

> link:../../experimental/sycl_ext_oneapi_kernel_arg_properties.asciidoc[sycl_ext_oneapi_kernel_arg_properties].

Author: I will remove alignment from the checked APIs because only cache control is what is being considered for implementation so far.
Alignments will be handled by the query like you suggested in your last comment. So users must make sure their pointers are 4 bytes aligned. Otherwise, results are undefined.

   const joint_matrix<Group, T, use::b, Rows, Cols, Layout> &res,
   ext::oneapi::experimental::annotated_ptr<T, PropertyListT> base_dest,
   size_t Stride, size_t GlobalRows, size_t GlobalCols,
   size_t RowIndex, size_t ColIndex);

} // namespace sycl::ext::intel::experimental::matrix
```

The property list associated with the `annotated_ptr` argument
represents the compile-time constant properties for cache control included
in the SYCL extenion
link:../../proposed/sycl_ext_intel_cache_controls.asciidoc[sycl_ext_intel_cache_controls].

==== Restrictions and Device Information Descriptors
Applications must adhere to certain alignment restrictions when using
the checked APIs described in this section. This extension provides
the following queries to get these requirements:

[frame="none",options="header"]
|======================
| Device descriptors | Return type| Description
|`ext::intel::experimental::info::device::matrix_checked_alignment`| `size_t`
|Tells the required alignment (in bytes) of the base pointer for
`joint_matrix_load_checked` and `joint_matrix_store_checked`.
|`ext::intel::experimental::info::device::matrix_checked_rowindex_multiple_of<T>`|

> `size_t`|Returns a value, of which `RowIndex` must be multiple of.

Reviewer 2: Kind of a nit, but it would be good to add a sentence:

> Where `T` is the element type of the matrix.

However, this makes me wonder how this works for a matrix that uses TF32.  Should `T` be `precision::tf32` or `float` (which is the storage type when using TF32)?

Author: We should be able to use the element type `precision::tf32`

Reviewer 2: I'd suggest adding another sentence to make this clear.  Something like:

> When using the matrices with the machine learning types, `T` should be the element type (e.g. `precision::tf32`) not the storage type.

|`ext::intel::experimental::info::device::matrix_checked_globalcols_multiple_of<T>`|
`size_t` | Returns a value, of which `GlobalCols` must be multiple of.
|======================

==== Appendix: Restrictions Per Hardware
===== Intel XMX
The checked APIs are currently available in devices with the architecture
`architecture::intel_gpu_pvc`. The following restrictions apply to
these checked APIs:

- The base pointer must be 4 bytes aligned.

- For 8 bits data type, `RowIndex` must be a multiple of 4. For 16 bits
data type, `RowIndex` must be a multiple of 2. So `RowIndex` must be a
multiple of 4 divided by size of the element type (`4/sizeof(T)`).

> - For 8 bits data type, `GlobalCols` must be a multiple of 4, For 16 bits

Reviewer 2: ```suggestion
- For 8 bits data type, `GlobalCols` must be a multiple of 4. For 16 bits
```

`col` represent the coordinates in the logical matrix before VNNI
transformation.

> === Load/Store/Fill With Out-of-Bounds Checks

Author: The check is done compared to the big matrix size comparing the address to be loaded (ptr) with the big matrix size (Height). If ptr > Height, the element should not be loaded.
if the remainder is passed, how is the implementation going to compare with it?

> === Load/Store/Fill With Out-of-Bounds Checks

Reviewer 2: With your recent change that adds `coordX` and `coordY` parameters, I see how this could work for the load and store APIs.  However, I still don't see how the fill API can work.

This API structure seems confusing to me, though, and probably less efficient.  I suspect users will get confused that the `src` pointer to `joint_matrix_load` is completely different from the `src_base` pointer to `joint_matrix_load_checked`.  It seems like it would be less confusing and more efficient to have the user pass values for `Width` and `Height` that are the *remaining* amount of memory in the big matrix.  Maybe this would be easier to explain in a meeting where we can draw pictures?

> === Load/Store/Fill With Out-of-Bounds Checks

Author: Yes, let's discuss fill and remainder approach in a meeting.

> |<------------`Stride`---------->|

Author: >> Using a horizontal line for "height" is confusing. 

This is not a horizontal line for "height". It is just one of the rows as "...". But I agree with the confusion. I will improve it.

>> "Stride" is normally a power of two?

There is no restriction on stride

I will match the numbers in the example with the picture

Note that joint matrix shape `Rows` and `Cols` represents a sub-block
of the picture above. The out of bounds data results when the global
matrix size is not evenly divisible by the joint matrix size.

==== Checked APIs
`joint_matrix_load`, `joint_matrix_store`, and `joint_matrix_fill`
operations do not perform bounds checking. When the global matrix size is
not multiple of the joint matrix size, this extension adds a new form
of API to load, store, and fill joint matrix while checking the
bounds. For load and fill, the out-of-bounds elements are set to

> 0. For the store, they are dropped out.

Author: Your description looks better, I will change to it. 
However, I don't prefer to use the word "padding" as this may suggest padding occurred (by the user or compiler) which is not necessarily the case, these data are just there in memory. So I would use the word "extra".

yes, the invalid address description is correct as these functions do not try to access the extra elements.

The property list associated with the `annotated_ptr` argument
represents the compile-time constant properties for cache control included
in the SYCL extenion
link:../../proposed/sycl_ext_intel_cache_controls.asciidoc[sycl_ext_intel_cache_controls].

==== Restrictions and Device Information Descriptors
The following restrictions apply to these checked APIs:

- The base pointer must be 4 bytes aligned.

- For 8 bits data type, `CoordX` must be a multiple of 4. For 16 bits
data type, `CoordX` must be a multiple of 2. So `CoordX` must be a
multiple of 4 divided by size of the element type (`4/sizeof(T)`).

- For 8 bits data type, `Width` must be a multiple of 4, For 16 bits
data type, `Width` must be a multiple of 2. So `Width` must be a

> multiple of 4 divided by size of the element type (`4/sizeof(T)`).

Reviewer 2: Describing it this way is confusing, and not really right.  For example, you first say that the base pointer must be 4 byte aligned, and then you say there is a query to get this alignment.  Presumably, we expect the alignment requirement to be different someday, so the spec should not say the requirement is 4 bytes.  I think you should remove these bullet points and say something like:

> Applications must adhere to certain alignment restrictions when using the checked APIs described in this section.  This extension provides the following queries to get these requirements:
>
> [Table]

The "Description" column will need to be expanded a bit to describe some of the details that are in the bullet points.  For example, the description of `matrix_checked_alignment` might say:

> Tells the required alignment (in bytes) of the base pointer for `joint_matrix_load_checked` and `joint_matrix_store_checked`.

If you want to document the alignment requirements of our current hardware, you could do that after the table describing the queries.  But you should somehow make it clear that the values only apply to our existing devices, and future devices may have different requirements.

Author: That's a good point, this is tying the API to a specific hardware which should be avoided.
yes, I will add the alignment requirements as an appendix like we did for the combinations table per each device. Here, these only apply to PVC for now.
